From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peter M <petermm@gmail.com>
Date: Sat, 24 Oct 2026 09:30:00 +0200
Subject: [PATCH] Add a riscv/esp32p4db port

Add lwIP support for ESP32-P4 (riscv/esp32p4db).
Mirrors the ESP32-S31 port with tuned memory footprint for internal SRAM.

Signed-off-by: Peter M <petermm@gmail.com>
---
 defs/bsps/riscv/esp32p4db.json  |  8 +++
 rtemslwip/esp32p4/lwipbspopts.h | 49 +++++++++++++++++++
 rtemslwip/esp32p4/netstart.c    | 84 ++++++++++++++++++++++++++++++++
 3 files changed, 141 insertions(+)
 create mode 100644 defs/bsps/riscv/esp32p4db.json
 create mode 100644 rtemslwip/esp32p4/lwipbspopts.h
 create mode 100644 rtemslwip/esp32p4/netstart.c

diff --git a/defs/bsps/riscv/esp32p4db.json b/defs/bsps/riscv/esp32p4db.json
new file mode 100644
index 0000000..8a23081
--- /dev/null
+++ b/defs/bsps/riscv/esp32p4db.json
@@ -0,0 +1,8 @@
+{
+	"header-paths-to-import": [
+		"rtemslwip/esp32p4"
+	],
+	"source-files-to-import": [
+		"rtemslwip/esp32p4/netstart.c"
+	]
+}
diff --git a/rtemslwip/esp32p4/lwipbspopts.h b/rtemslwip/esp32p4/lwipbspopts.h
new file mode 100644
index 0000000..f6e40f3
--- /dev/null
+++ b/rtemslwip/esp32p4/lwipbspopts.h
@@ -0,0 +1,49 @@
+/*
+ * lwIP sizing for riscv/esp32p4db.
+ *
+ * Copyright (C) 2026 Peter M. <petermm@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#ifndef RTEMSLWIP_ESP32P4_LWIPBSPOPTS_H
+#define RTEMSLWIP_ESP32P4_LWIPBSPOPTS_H
+
+/*
+ * 8 KiB heap for lwIP dynamic allocations.
+ */
+#define MEM_SIZE           ( 8 * 1024 )
+
+/*
+ * PBUF pool sizing: 8 pbufs of 1600 bytes.
+ */
+#define PBUF_POOL_SIZE      8
+#define PBUF_POOL_BUFSIZE  1600
+
+#define MEMP_NUM_PBUF      16
+#define MEMP_NUM_RAW_PCB    4
+#define MEMP_NUM_UDP_PCB    6
+#define MEMP_NUM_TCP_PCB    8
+#define MEMP_NUM_TCP_PCB_LISTEN 4
+#define MEMP_NUM_TCP_SEG   24
+#define MEMP_NUM_NETBUF     8
+#define MEMP_NUM_NETCONN   10
+#define MEMP_NUM_SELECT_CB  4
+#define MEMP_NUM_TCPIP_MSG_API 8
+#define MEMP_NUM_TCPIP_MSG_INPKT 16
+
+#define TCPIP_MBOX_SIZE    16
+#define DEFAULT_RAW_RECVMBOX_SIZE 8
+#define DEFAULT_UDP_RECVMBOX_SIZE 8
+#define DEFAULT_TCP_RECVMBOX_SIZE 8
+#define DEFAULT_ACCEPTMBOX_SIZE   4
+
+#define TCP_MSS            1460
+#define TCP_SND_BUF        ( 4 * TCP_MSS )
+#define TCP_WND            ( 4 * TCP_MSS )
+#define TCP_SND_QUEUELEN   ( ( 4 * TCP_SND_BUF ) / TCP_MSS )
+
+#define LWIP_NETIF_STATUS_CALLBACK  1
+
+#endif /* RTEMSLWIP_ESP32P4_LWIPBSPOPTS_H */
diff --git a/rtemslwip/esp32p4/netstart.c b/rtemslwip/esp32p4/netstart.c
new file mode 100644
index 0000000..21d9f98
--- /dev/null
+++ b/rtemslwip/esp32p4/netstart.c
@@ -0,0 +1,84 @@
+/*
+ * start_networking() for riscv/esp32p4db.
+ *
+ * Copyright (C) 2026 Peter M. <petermm@gmail.com>
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials publish with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES ARE DISCLAIMED.
+ */
+
+#include <netstart.h>
+
+#include <lwip/netif.h>
+#include <lwip/tcpip.h>
+
+#include <rtems/bspIo.h>
+
+int __attribute__((weak)) esp32_netif_add(
+  struct netif  *net_interface,
+  ip_addr_t     *ipaddr,
+  ip_addr_t     *netmask,
+  ip_addr_t     *gateway,
+  unsigned char *mac_address
+);
+
+int __attribute__((weak)) esp32p4_netif_add(
+  struct netif  *net_interface,
+  ip_addr_t     *ipaddr,
+  ip_addr_t     *netmask,
+  ip_addr_t     *gateway,
+  unsigned char *mac_address
+)
+{
+  if (esp32_netif_add != NULL) {
+    return esp32_netif_add(net_interface, ipaddr, netmask, gateway, mac_address);
+  }
+
+  (void) net_interface;
+  (void) ipaddr;
+  (void) netmask;
+  (void) gateway;
+  (void) mac_address;
+
+  printk(
+    "esp32p4: the stack is up but there is no interface. Link a driver providing\n"
+    "esp32p4: esp32p4_netif_add() or esp32_netif_add() to get one.\n"
+  );
+
+  return 1;
+}
+
+int start_networking(
+  struct netif  *net_interface,
+  ip_addr_t     *ipaddr,
+  ip_addr_t     *netmask,
+  ip_addr_t     *gateway,
+  unsigned char *mac_address
+)
+{
+  int ret;
+
+  ret = start_networking_shared();
+  if ( ret != 0 ) {
+    return ret;
+  }
+
+  return esp32p4_netif_add(
+    net_interface,
+    ipaddr,
+    netmask,
+    gateway,
+    mac_address
+  );
+}
-- 
2.50.1 (Apple Git-155)
